auth/couchdb_session_authenticator.go

```diff
@@ -116,23 +115,30 @@ func (a *CouchDbSessionAuthenticator) Validate() error {
 
 // Authenticate adds session authentication cookie to a request.
 func (a *CouchDbSessionAuthenticator) Authenticate(request *http.Request) error {
-
 	a.URL = request.URL.Scheme + "://" + request.URL.Host
 	a.header = request.Header
 	a.ctx = request.Context()
 
-	cookie, err := a.getCookie()
-	if err != nil {
-		return err
+	err := a.refreshCookie()
+
+	if a.client.Jar == nil && a.session != nil {
+		if cookie := a.session.getCookie(); cookie != nil {
+			request.AddCookie(cookie)
+		}
 	}
 
-	request.AddCookie(cookie)
-	return nil
+	return err
 }
 
-// getCookie returns an AuthSession cookie to be used in a request.
+// SetClient sets the http client for the authenticator. It injects a cookiejar
+// if supplied client is missing one.
+func (a *CouchDbSessionAuthenticator) SetClient(client *http.Client) {
```

Comment:
I think this function should set the "inherited" value `DisableSSLVerification`.

Actually thinking about it more that is a property of the "service" and not the `http.Client` so this isn't the correct place to read the new value and set it. Another complication is that the `URL` is also inhereted. Currently we declare:
>// CouchDB URL inherited from the service config.
and
>// A flag that indicates whether verification of the server's SSL certificate should be disabled; INherired from the service config

Looking at the code I think that was possibly only ever true if the authenticator was init'd from `GetAuthenticatorFromEnvironment` and never reflected any subsequent changes caused by calls to either `SetHttpClient` or `SetServiceURL`. I think this remains true at the current state of this PR.

I don't see there is any value now in exposing `DisableSSLVerfication` as it will always be doing whatever the `SetClient` client has been configured to do (and that in turn will pretty much always be the base service client).
`URL` is a little more complicated because we will need access to set that. So we can either leave it exposed or internalize it as well and add a `SetURL` similar to `SetClient`. In either case the value will need to be updated from the `SetServiceURL` override.

Reply:
I have a commit ["fix: put back apparently required auth field"](https://github.com/IBM/cloudant-go-sdk/pull/399/commits/0cc08ba21762bb454fe5daa80f1eded1f1c44b23) that says it all 😄 

This is basically part of core's support for `PROPNAME_SVC_DISABLE_SSL` and we use it as a public var in our base service override, so can't be changed or removed.

> I think this function should set the "inherited" value DisableSSLVerification.

I'm not so sure. I used to set it on constructed http client, because it could come from env, but now we are getting this for free from base service and are treating `SetClient` as "do whatever you want" function. It'd be weird to override one arbitrary parameter in programmatical setup based on value from env.

It probably should be used in `NewCouchDbSessionAuthenticator` directly instead, but given that we are almost immediately throwing that client away I feel this is more hustle than it worth it. 


`URL` is more complicated, because it is a property of request, not http client and we reconstruct it from request's URL evey time.

Basically both `URL` and `DisableSSLVerification`  are part of Authenticator interface, though implicit ones, rather than real meaningful vars of session authenticator.

Reply:
Alright, I can live with the disable SSL stcking around and the flag potentially being inconsistent with the current client (my hatred of that particular option is well known and anyone using it has bigger problems than this IMO)!

I'm ok with `URL` as public, but I do still think that `SetServiceURL` should update the session authenticator URL. If I follow the pattern to get a new default instance (which points to `http://localhost:5984`) and then call:
`service.SetServiceURL('https://myservice.example')` I'd want it to be trying to auth against `https://myservice.example/_session` not `https://localhost:5984/_session`.

Reply:
Well, that's not a concern, because we are [composing URL](https://github.com/IBM/cloudant-go-sdk/blob/main/auth/couchdb_session_authenticator.go#L120) from request on each authenticate call, so `auth.URL` is more of a container to consequent calls to `_session` after that, it's never going to be out of sync. Honestly we have it as public for the same reason as `DisableSSLVerification` - those are "hidden" authenticator interface expectations.

I did this anyway in https://github.com/IBM/cloudant-go-sdk/pull/399/commits/0749c50ba1d8f962b63173838b500c19777245be just for completion sake.

Reply:
Ah a subtle deviation from what the other implementations do that I had overlooked as one of the pieces that isn't actually changing here!
Thanks for closing the loop though.
I guess the per request `URL` code isn't strictly required any more, but I think it can stay (particularly to help in those weird potential 2 client scenarios we've discused).

Reply:
Exactly. We can remove it now, just set auth's URL on client inject and on `SetServiceURL`, but then that "you can set your client" scenario is becoming a bit more awkward, because user will additionally have to set URL too or it'll stick to whatever was set there by default. On the other hand this is what user is probably want,  so I can't decide one way or another.